src/metadata/metadata.rs

```diff
@@ -0,0 +1,207 @@
+            "Abstract" => {
+                let val = parse::parse_vec(val);
+                self.abs.join(Some(val));
+            }
+            "Canonical Url" => {
+                let val = val.clone();
+                self.canonical_url.join(Some(val));
+            }
+            "Date" => {
+                let val = match parse::parse_date(val) {
+                    Ok(val) => val,
+                    Err(_) => {
+                        die!("The \"Date\" field must be in the format YYYY-MM-DD."; line_num)
+                    }
+                };
+                self.date.join(Some(val));
+            }
+            "ED" => {
+                let val = val.clone();
+                self.ed.join(Some(val));
+            }
+            "Editor" => {
+                let val = parse::parse_editor(val);
+                self.editors.join(Some(val));
+            }
+            "Group" => {
+                let val = val.clone();
+                self.group.join(Some(val));
+            }
+            "Level" => {
+                let val = parse::parse_level(val);
+                self.level.join(Some(val));
+            }
+            "Shortname" => {
+                let val = val.clone();
+                self.shortname.join(Some(val))
+            }
+            "Status" => {
+                let val = val.clone();
+                self.raw_status.join(Some(val));
+            }
+            "Title" => {
+                let val = val.clone();
+                self.title.join(Some(val));
+            }
+            _ => die!("Unknown metadata key \"{}\".", key; line_num),
+        }
+
+        self.has_metadata = true;
+    }
+
+    pub fn fill_macros(&self, doc: &mut Spec) {
+        let macros = &mut doc.macros;
+
+        if let Some(date) = self.date.as_ref() {
+            macros.insert(
+                "date",
+                date.format(&format!("{} %B %Y", date.day())).to_string(),
+            );
+            macros.insert("isodate", date.to_string());
+        }
+        if let Some(level) = self.level.as_ref() {
+            macros.insert("level", level.clone());
+        }
+        if let Some(shortname) = self.shortname.as_ref() {
+            macros.insert("shortname", shortname.clone());
+        }
+        if let Some(raw_status) = self.raw_status.as_ref() {
+            macros.insert(
+                "longstatus",
+                SHORT_TO_LONG_STATUS
+                    .get(raw_status.as_str())
+                    .unwrap()
+                    .to_string(),
+            );
+        }
+        if let Some(title) = self.title.as_ref() {
+            macros.insert("title", title.clone());
+            macros.insert("spectitle", title.clone());
+        }
+    }
+
+    pub fn compute_implicit_metadata(&mut self) {
+        if self.canonical_url.is_none() || self.canonical_url.as_ref().unwrap() == "ED" {
+            self.canonical_url = self.ed.clone();
+        }
+    }
+
+    pub fn validate(&self) {
```

Comment:
This can just be an assert in `parse_metadata` instead.

Reply:
Putting an assert in `parse_metadata` would only check if there's any metadata in the input file, but we can provide metadata in the command line without giving that in the input file, so checking metadata in `validate` would be better.

Reply:
Well, then the assertion message is wrong, right? If it wants to assert that there's a `<pre class=metadata>` block then it should just assert in parse_metadata. Otherwise yeah we need a separate thing.